dbms/src/Storages/DeltaMerge/Delta/DeltaValueSpace.h

```diff
@@ -94,7 +94,8 @@ class DeltaValueSpace
     DeltaIndexPtr delta_index;
 
     // Protects the operations in this instance.
-    mutable std::mutex mutex;
+    // It is a recursive_mutex because the lock may be also used by the parent segment as its update lock.
+    mutable std::recursive_mutex mutex;
```

Comment:
Note: This recursive mutex happens when:

1. DeltaMergeStore::segmentIngestData grabs a segment update lock
2. DeltaMergeStore::segmentIngestData discovered the segment is not empty and then ingest files into the delta.
   - DeltaValueSpace::ingestColumnFiles grabs the same lock to write to delta VS.

Reply:
Usually, in `DeltaMergeStore::segmentApplyXXX` we will acquire the segment update lock and won't change the DeltaVS inside the "apply".
`segmentIngestData` brings a special pattern that may change the DeltaVS inside its "apply". But I didn't figure out a better solution